### PR TITLE
Maintenance: Removed most NULLs using modernize-use-nullptr

### PR DESCRIPTION
**Pull request by rousskov**
Applying clang-tidy modernize-use-nullptr also converted many
"anonymous" 0s into nullptr.

Clang-tidy uses clang AST, so we can only modernize code what clang can
compile. Fortunately, it is fairly easy to compile most of Squid code,
eliminating ~90% of NULLs in actively maintained code (excluding
comments and Windows-specific code) and ~73% of all NULLs.

TODO: Consider eliminating all remaining C++ NULLs using a simple script
that does not understand C++ syntax, assuming that the remaining cases
are simple enough to avoid problems that a compiler cannot detect.

### CODE REVIEW

**Comment by rousskov** [comment]
A slightly simplified version of the script that produced these changes is [available](https://gist.github.com/rousskov/58183e672749ac7f9f3c22df1420c801). If this master/v6 PR is merged, I would be happy to apply similar conversion to the official v5 branch so that we can keep the two branches in sync, reducing merge conflicts.

Very approximate and unaudited statistics:

Before:

```
All NULLs: 5570
C NULLs:   592
C++ NULLs: 4904
C++ NULLs outside of comments and Windows code: 4529

Top contributors:
src/acl/external/kerberos_ldap_group/support_ldap.cc:116
src/acl/external/eDirectory_userip/ext_edirectory_userip_acl.cc:94
src/cache_cf.cc:85
src/client_side.cc:71
src/snmp_core.cc:68
src/esi/Esi.cc:68
src/neighbors.cc:61
src/clients/FtpGateway.cc:60
src/dns_internal.cc:55
src/auth/UserRequest.cc:55
```

After:
```
All NULLs: 1482
C NULLs:   592
C++ NULLs: 816
C++ NULLs outside of comments and Windows code: 449

Top contributors:
src/acl/external/LM_group/ext_lm_group_acl.cc:46
src/acl/external/AD_group/ext_ad_group_acl.cc:46
src/auth/ntlm/SMB_LM/ntlm_smb_lm_auth.cc:44
src/dns_internal.cc:23
src/acl/external/kerberos_ldap_group/support_ldap.cc:23
src/auth/ntlm/SSPI/ntlm_sspi_auth.cc:19
src/comm/ModDevPoll.cc:12
src/test_cache_digest.cc:11
src/auth/basic/PAM/basic_pam_auth.cc:10
src/globals.h:9
```

**Review by kinkie — APPROVED** [review]
I have sample-reviewed several files, and it looks good to me, it improves clarity and robustness.
It highlights a lot of "pointer == nullptr" or "pointer != nullptr" idioms that feel a bit weird, but done is better than perfect so I'd land it and worry about them later on

File: src/Transients.cc [hunk 1]
```diff
@@ -49,7 +49,7 @@ Transients::init()
     map->cleaner = this;
     map->disableHitValidation(); // Transients lacks slices to validate
 
-    locals = new Locals(entryLimit, 0);
+    locals = new Locals(entryLimit, nullptr);
```
Comment (rousskov) [inline src/Transients.cc:52]: This is one of the examples where clang-tidy was able to replace `0` with nullptr.

File: include/splay.h [hunk 2]
```diff
@@ -153,13 +153,13 @@ SplayNode<V>::remove(Value const dataToRemove, SPLAYCMP * compare)
     if (splayLastResult == 0) { /* found it */
         SplayNode<V> *newTop;
 
-        if (result->left == NULL) {
+        if (result->left == nullptr) {
```
Comment (rousskov) [inline include/splay.h:156]: > [Francesco](https://github.com/squid-cache/squid/pull/1075#pullrequestreview-1022406459): It highlights a lot of "pointer == nullptr" or "pointer != nullptr" idioms that feel a bit weird, but done is better than perfect so I'd land it and worry about them later on

I agree. And there were dozens of such bad expressions before this PR (almost 2000 if you count their NULL versions!). We are not making things _worse_ here by introducing something that did not exist before...

Some day, we will find and use a tool that will safely eliminate these expressions, but I doubt we should wait (may take too long while removing NULLs has immediate and long-awaited benefits) or do it now with a dumb script (too dangerous?). Said that, if others are sure that we should remove most of such expressions in this PR (to reduce the number of code changes over time), then I will adjust the PR to do it (with a dumb script).

Comment (kinkie) [reply include/splay.h:156]: I am happy with this to go in, as highlighted by me approving the PR

**Comment by rousskov** [comment]
@yadij, just FYI: I plan to clear this PR for merging in three days unless you stop me.

**Pull request closed without merging**